File: 2026/01/pascal-triangle/pascal-triangle1.py
def generate_spaces(count):
    return '  ' * count


def draw_pt(level):
    # validate input
    if level < 0 or not isinstance(level, int):
        return
    
    for i in range(level + 1):
        prev = 1
        print(generate_spaces(level - i), end='')
        for j in range(i + 1):
            print(f'{prev}   ', end='')
            # calculate next number
            prev = prev * (i - j) // (j + 1)
        print()


def draw_sierpinski(level):
    # validate input
    if level < 0 or not isinstance(level, int):
        return
    
    for i in range(level + 1):
        prev = 1
        print(generate_spaces(level - i), end='')
        for j in range(i + 1):
            # 'Q' for odd, '.' for even
            symbol = 'Q' if prev % 2 else '.'
            print(f'{symbol}   ', end='')
            # calculate next number
            prev = prev * (i - j) // (j + 1)
        print()


if __name__ == '__main__':
    draw_sierpinski(54)

File: 2026/01/pascal-triangle/pascal-triangle1.ts
const generateStringWithSpaces = (count: number): string => {
    return '  '.repeat(count);
  }

export function drawPT(level: number): void {
    if (level < 0 || Number(level) !== level) {
        return;
    }

    for (let i = 0; i <= level; i++) {
        let prev = 1;
        let row = '';
        process.stdout.write(generateStringWithSpaces(level - i));
        for (let j = 0; j <= i; j++) {

           process.stdout.write(`${prev}`);
            prev = prev * (i - j) / (j + 1); 
            

           
        }
      process.stdout.write(`\n`);
    }

}

export function drawSierpinski(level: number): void {
    if (level < 0 || Number(level) !== level) {
        return;
    }

    for (let i = 0; i <= level; i++) {
        let prev = 1;
        let row = '';
        process.stdout.write(generateStringWithSpaces(level - i));
        for (let j = 0; j <= i; j++) {
            // [i, j]
           // console.log(prev);
           process.stdout.write(`${prev % 2 ? 'Q' : '.'}   `);
            prev = prev * (i - j) / (j + 1); 
            

           
        }
      process.stdout.write(`\n`);
    }

}

drawSierpinski(52);
